lib/src/main/java/org/esbtools/lightbluenotificationhook/NotificationHook.java

```diff
@@ -88,7 +96,9 @@ public void processHook(EntityMetadata entityMetadata,
         Mediator mediator = tryGetMediator();
 
         Projector watchProjector = Projector.getInstance(config.watchProjection(), entityMetadata);
-        Projector includeProjector = Projector.getInstance(config.includeProjection(), entityMetadata);
+        Projection includeProjection=addArrayIdentities(config.includeProjection(),entityMetadata);
+        Projector includeProjector = Projector.getInstance(includeProjection, entityMetadata);
```

Comment:
You are right. I started doing this with the correct intent, but got
distracted with something unrelated for a while, and when I got back, I
ended up with this. Context switch failure. I'll fix.

On Thu, Mar 17, 2016 at 3:07 PM, Alec Henninger <notifications@github.com>
wrote:

> In
> lib/src/main/java/org/esbtools/lightbluenotificationhook/NotificationHook.java
> <https://github.com/esbtools/lightblue-notification-hook/pull/10#discussion_r56577875>
> :
>
> > @@ -88,7 +96,9 @@ public void processHook(EntityMetadata entityMetadata,
> >          Mediator mediator = tryGetMediator();
> >
> >          Projector watchProjector = Projector.getInstance(config.watchProjection(), entityMetadata);
> > -        Projector includeProjector = Projector.getInstance(config.includeProjection(), entityMetadata);
> > +        Projection includeProjection=addArrayIdentities(config.includeProjection(),entityMetadata);
> > +        Projector includeProjector = Projector.getInstance(includeProjection, entityMetadata);
>
> Not sure, but I think we may have wanted to modify the watch projection
> instead of the include projection.
>
> Here is a test case which demonstrates the bug I was running into. In this
> PR branch it does not pass (though double check me to make sure I don't
> have something wrong):
>
>     @Test
>     public void shouldOnlyIncludeWhatChangedOfArrayElementIfElementIdentityNotWatched()
>             throws Exception {
>         EntityMetadata md = getMd("usermd.json");
>         JsonNode pre = loadJsonNode("userdata.json");
>         JsonNode post = loadJsonNode("userdata.json");
>         JsonDoc.modify(post, new Path("sites.0.streetAddress.city"), JsonNodeFactory.instance.textNode("blah"), true);
>
>         List<HookDoc> docs= new ArrayList<>();
>         HookDoc doc = new HookDoc(md, new JsonDoc(pre), new JsonDoc(post), CRUDOperation.UPDATE, "me");
>         docs.add(doc);
>
>         HookConfiguration cfg = new NotificationHookConfiguration(
>                 projection("{'field':'sites.*.streetAddress.city','recursive':1}"),
>                 null,
>                 false);
>
>         hook.processHook(md, cfg, docs);
>         Assert.assertNotNull(insertCapturingMediator.capturedInsert);
>         JsonNode data=insertCapturingMediator.capturedInsert.getEntityData();
>
>         Truth.assertThat(
>                 Iterables.transform(data.get("changedPaths"), toTextValue()))
>                 .containsExactly("sites.0.streetAddress.city");
>     }
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/esbtools/lightblue-notification-hook/pull/10/files/0507ae1f8980cdc0d11725e3066130cbf63a5604#r56577875>
>